Make cycle detection throw an exception instead of returning null.

Danieladu opened the pull request:
Fixes #4490

## Description
Remove the two-step passes and the cache logic. Just throw an exception when detecting the loop.

carlosscastro commented:
Thanks for working on this. I'm not sure that this is the route we want to go. There are customers that require usage of cycles. On the other hand, we know that PVA wants to be able to jsut throw an exception when cycles are found, so we might want to configure 1) throw on cycles or 2) load graph with cycles. However the cycle should not return null -- do you ahve an example graph that results in null?

Danieladu commented:
> Thanks for working on this. I'm not sure that this is the route we want to go. There are customers that require usage of cycles. On the other hand, we know that PVA wants to be able to jsut throw an exception when cycles are found, so we might want to configure 1) throw on cycles or 2) load graph with cycles. However the cycle should not return null -- do you ahve an example graph that results in null? If there is a bug in cycle detection, then there are 2 things to do here: first, make it configurable and second, fix the bug if there is one. If there is a bug, feel free to share the json and conversation that triggers the bug and I could help.

Thanks for the reply.
I have no idea that the issue filed by tom means we should make it configurable or just throw an exception. (Maybe throw directly is a breaking change for some customers.)
The question is:
1. Configurable or just throw
2. if configurable, the default config is throwing or allow the cycles.
@tomlm  can you give the goal of this issue? Thanks.

boydc2014 commented:
Thanks @Danieladu. Based on @carlosscastro's message above, since some customers is depending on this to not throw, so making it configurable makes sense to me, we definitely prefer not to break any existing scenario. Thus make the default value to "not throw" make sense, because it will achieve the best back-ward compatible.

Danieladu commented:
hi @carlosscastro, I have made "allowCycle" configurable.  Please take a look.

boydc2014 commented on libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs:
```diff
@@ -41,8 +42,10 @@ public class ResourceExplorer : IDisposable
         /// Initializes a new instance of the <see cref="ResourceExplorer"/> class.
         /// </summary>
-        public ResourceExplorer()
+        /// <param name="allowCycle">If allowCycle is set to true, throw an exception when detecting cycle.</param>
```
if allowCycle is false?

carlosscastro commented:
Thanks guys, you got it right, this way customers can choose their behavior. Appreciate it!

carlosscastro commented on libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Memory/Expressions/Converters/DialogExpressionConverter.cs:
```diff
@@ -57,9 +57,14 @@ public override DialogExpression ReadJson(JsonReader reader, Type objectType, Di
+                    catch (Exception e)
 #pragma warning restore CA1031 // Do not catch general exception types
                     {
+                        if (e.Message.StartsWith("Cycle detected", StringComparison.InvariantCulture))
```
Intead of this approach which is a bit brittle, we could throw a more specific exception in the cycle, and add a new catch on top. For example

```csharp
catch (InvalidOperationException e)
{
    throw;
}
catch (Exception e)
{
    result = new DialogExpression($"='{id}'");
}
```

I'm not sure what the right exception is, I'll put a bit of thought into it, but feel free to propose. Ideally no custom exceptions. Another approach could be fine also, as long as we don't parse the exception text.

carlosscastro commented on libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Observers/CycleDetectionObserver.cs:
```diff
@@ -49,6 +59,11 @@ public bool OnBeforeLoadToken<T>(SourceContext context, SourceRange range, JToke
 
+            if (isCycle && !allowCycle)
+            {
+                throw new Exception($"Cycle detected for range: {range}");
```
Consider throwing a more specific exception type.

carlosscastro commented:
The changes are going in the right direction! I added comments for some details. Let me know if anything doesn't make sense to you. Ping me on Teams if you want to chat or I can be of help. thanks!

Danieladu commented:
> The changes are going in the right direction! I added comments for some details. Let me know if anything doesn't make sense to you. Ping me on Teams if you want to chat or I can be of help. thanks!

The suggestion is very useful. I have changed the common Exception type to a specific type.

carlosscastro commented on tests/Microsoft.Bot.Builder.Dialogs.Declarative.Tests/JsonLoadTests.cs:
```diff
@@ -60,6 +65,12 @@ await BuildTestFlow(@"Root.dialog")
+        [TestMethod]
+        public void JsonDialogLoad_CycleDetectionWithNoCycleMode()
+        {
+            Assert.ThrowsException<Exception>(() => BuildNoCycleTestFlow(@"Root.dialog"));
```
Don't we want to catch InvalidOperationException here? Even though the current assertion is true, a more specific assertion would be better.

```suggestion
            Assert.ThrowsException<InvalidOperationException>(() => BuildNoCycleTestFlow(@"Root.dialog"));
```

carlosscastro commented on libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs:
```diff
@@ -41,8 +42,10 @@ public class ResourceExplorer : IDisposable
         /// </summary>
-        public ResourceExplorer()
+        /// <param name="allowCycle">If allowCycle is set to true, throw an exception when detecting cycle.</param>
+        public ResourceExplorer(bool allowCycle = true)
```
The PR is looking good overall. The only thing that doesn't feel great is having the `allowCycle` as a ResourceExplorer constructor parameter. I'll think of alternatives to this today, to see if some feels more correct. I do understand that whatever they set needs to be injectable through DI which is why it was added here. @tomlm fyi.

carlosscastro replied:
I enlisted on this branch and tried a few things to avoid this parameter being passed through the resource explorer but I didn't find a way without significant refactors. The main reason, in my opinion, is that the resource explorer should focus on resources, and there should be a separate abstraction for type / json loading. That abstraction should receive the allowCycle and not the resource explorer. The resource explorer should just interact (back and forth) with the typeloading abstraction. So this is in a way a problem with this class having too many responsibilities. 

Anyways, I'm proposing some potential refactors this week for longer term with the SDK team that is taking over this code, but for now we can go with this.

carlosscastro replied:
We discussed offline that as a temporary measure we'll have a ResourceExplorerOptions to avoid having so specific parameters and allow extensibility until a larger refactor takes place.

carlosscastro commented:
@Danieladu and I jsut discussed a small change where `ResourceExplorer` will receive a `REsourceExplorerOptions` object that includes an `AllowCycle` bool property. Please don't merge until then.
